### PR TITLE
Put Permutation-related Proper instances at default priority.

### LINKED ISSUE #14571: Permutation_app' has wrong priority
#### Description of the problem

Why is the priority of `Permutation_app' : Proper (Permutation ==> Permutation ==> Permutation) app` level 10? This is higher than the priority 5 of `proper_subrelation` in `Morphisms`, which means that when looking for `Proper (_ ==> _ ==> _) app` with one of the underscores being `Permutation` the tactic `proper_subrelation` is triggered before the instance `Permutation_app'` is used. As shown in https://gitlab.mpi-sws.org/iris/stdpp/-/issues/114, this can result in a blow-up in instance search during setoid rewriting:

```coq
From stdpp Require Import gmultiset.

Typeclasses eauto := debug.

Lemma test {A B} `{Countable A} `{Countable B}
      (xs : list (A * list B)) (X Y: gmultiset (A * list B)) :
  xs ++ elements (X ⊎ Y) ≡ₚ xs ++ elements X ++ elements Y.
Proof.
  Time rewrite gmultiset_elements_disj_union.
  (* Very slow: takes like 1 sec, produces a type class trace of 85.000 lines *)
Abort.

(** Remove this instance, since its priority is 10, which is above the priority 5 of [proper_relation] *)
Global Remove Hints Permutation_app' : typeclass_instances.
(** Add it with the right priority (default = 0, since the instance has no premises) *)
Existing Instance Permutation_app'.

Lemma test {A B} `{Countable A} `{Countable B}
      (xs : list (A * list B)) (X Y: gmultiset (A * list B)) :
  xs ++ elements (X ⊎ Y) ≡ₚ xs ++ elements X ++ elements Y.
Proof.
  Time rewrite gmultiset_elements_disj_union.
  (* Immediate: takes like 0.02 sec, produces a type class trace of 40 lines *)
Abort.
```

#### Coq Version

```
$ coqc --version 
The Coq Proof Assistant, version 8.13.1 (March 2021)
compiled on Mar 13 2021 10:35:46 with OCaml 4.11.1
```

The above example uses std++ dev.2021-06-28.1.ddce76d7. (With newer versions of std++ this problem might disappear since we might repair the priority of the instance in std++ until it's fixed in all versions of Coq that we support).

### PR DESCRIPTION
**Pull request by SkySkimmer**
Instances for `Proper (_ ==> eq)` have been left at previous priority.

Fixes #14571

### CODE REVIEW

**Comment by SkySkimmer** [comment]
cc @robbertkrebbers

**Comment by SkySkimmer** [comment]
Bench https://gitlab.com/coq/coq/-/jobs/1385217599

**Comment by SkySkimmer** [comment]
~~~
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│        coq-mathcomp-algebra │  140.94   141.68  -0.52 │   392239917267    393015974759  -0.20 │   511022835602    510954912888   0.01 │  554044   553888   0.03 │   0    0    nan │
│  coq-performance-tests-lite │ 1551.45  1557.70  -0.40 │  4332630213034   4349810465686  -0.39 │  7128465143259   7127491806675   0.01 │ 2079328  2079308   0.00 │   0    0    nan │
│       coq-mathcomp-fingroup │   48.15    48.34  -0.39 │   133373649234    133320336117   0.04 │   186295199764    186204403584   0.05 │  481976   482140  -0.03 │   0    0    nan │
│      coq-mathcomp-ssreflect │   53.74    53.90  -0.30 │   146817446003    146774256163   0.03 │   190182275428    190195514149  -0.01 │  526484   526340   0.03 │   0    0    nan │
│             coq-lambda-rust │ 1264.33  1267.83  -0.28 │  3520615572792   3525061467512  -0.13 │  5074819874269   5070724072526   0.08 │ 1096900  1097176  -0.03 │   0    0    nan │
│          coq-mathcomp-field │  222.36   222.71  -0.16 │   619181970168    619085598799   0.02 │   940093203469    940160919531  -0.01 │  668912   669032  -0.02 │   0    0    nan │
│            coq-math-classes │  174.15   174.40  -0.14 │   485569330598    486161339444  -0.12 │   644035970251    644058560738  -0.00 │  481180   481216  -0.01 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  736.19   736.92  -0.10 │  2050127610383   2052889424012  -0.13 │  3213490140086   3213144406644   0.01 │ 1071772  1071876  -0.01 │   0    0    nan │
│              coq-verdi-raft │ 1103.59  1104.66  -0.10 │  3074191696758   3075908129772  -0.06 │  4324217038436   4323314260005   0.02 │  902824   902924  -0.01 │   0    0    nan │
│               coq-perennial │ 6827.90  6833.49  -0.08 │ 19014520014315  19034167412217  -0.10 │ 29076332425956  29108517064510  -0.11 │ 3090156  3083744   0.21 │   0    0    nan │
│                   coq-color │  483.63   484.00  -0.08 │  1349904348669   1349280347437   0.05 │  1699938525429   1698671688454   0.07 │ 1153636  1153096   0.05 │   0    0    nan │
│                  coq-geocoq │ 1338.62  1339.22  -0.04 │  3734181575228   3733588712156   0.02 │  5469080630582   5468095829876   0.02 │  983232   983208   0.00 │   0    0    nan │
│       coq-engine-bench-lite │  312.36   312.46  -0.03 │   878535956791    880325270497  -0.20 │  1638103272001   1640233814117  -0.13 │ 3506172  3506172   0.00 │   0    0    nan │
│                coq-bedrock2 │  311.47   311.45   0.01 │   863466761284    861750558339   0.20 │  1506056371704   1505270054343   0.05 │  770936   770904   0.00 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1055.92  1055.16   0.07 │  2939487247644   2939954893227  -0.02 │  4665738825147   4665765833271  -0.00 │  937184   904084   3.66 │   0    0    nan │
│             coq-fiat-crypto │ 4811.45  4807.75   0.08 │ 13363684780579  13358119129984   0.04 │ 23580205310055  23581364300475  -0.00 │ 2222216  2220976   0.06 │   0    0    nan │
│               coq-fourcolor │ 2552.45  2550.33   0.08 │  7116217107425   7111183636680   0.07 │ 12238355624896  12239334618618  -0.01 │  730996   731044  -0.01 │   0    0    nan │
│            coq-fiat-parsers │  582.90   582.37   0.09 │  1629672699739   1629880592876  -0.01 │  2496502486638   2495925911528   0.02 │ 3475488  3475228   0.01 │   0    0    nan │
│                 coq-unimath │ 4264.62  4260.49   0.10 │ 11867773151466  11858126136582   0.08 │ 22673272276722  22682841952324  -0.04 │ 3262980  3283732  -0.63 │   0    0    nan │
│                   coq-flocq │  146.62   146.47   0.10 │   407196163729    407100306206   0.02 │   504461522074    504481426499  -0.00 │  808248   808192   0.01 │   0    0    nan │
│       coq-mathcomp-solvable │  170.06   169.82   0.14 │   472860766842    472849287046   0.00 │   668244423688    668243790946   0.00 │  695164   673100   3.28 │   0    0    nan │
│                coq-coqprime │  328.79   328.25   0.16 │   912988350314    912193798471   0.09 │  1595681779818   1594856157124   0.05 │  920168   919580   0.06 │   0    0    nan │
│                  coq-stdlib │  414.64   413.95   0.17 │  1159135951778   1157614954000   0.13 │  1474138004791   1474297716252  -0.01 │  594776   594736   0.01 │   0    0    nan │
│      coq-mathcomp-character │  150.44   150.17   0.18 │   417833740890    417660331084   0.04 │   584828154493    584847235642  -0.00 │  720076   719740   0.05 │   0    0    nan │
│                 coq-coqutil │   59.16    58.90   0.44 │   163496087273    163059291639   0.27 │   214626322176    214095656974   0.25 │  602368   602324   0.01 │   0    0    nan │
│                    coq-corn │ 1474.42  1467.54   0.47 │  4102483814176   4095019684891   0.18 │  5898293017766   5896198611784   0.04 │  813112   812792   0.04 │   0    0    nan │
│                    coq-hott │  280.84   279.42   0.51 │   770910833390    770191522736   0.09 │  1191651103858   1191541122388   0.01 │  569552   569620  -0.01 │   0    0    nan │
│                coq-rewriter │  671.40   667.78   0.54 │  1862854346695   1853886314641   0.48 │  2845911868453   2843710472706   0.08 │ 1071940  1071976  -0.00 │   0    0    nan │
│               coq-fiat-core │  108.80   107.82   0.91 │   306015715224    304984887983   0.34 │   413832141355    412397963779   0.35 │  480968   480944   0.00 │   0    0    nan │
│                   coq-verdi │   99.42    98.42   1.02 │   273319878028    270473642581   1.05 │   379911286846    378835443319   0.28 │  543868   544060  -0.04 │   0    0    nan │
│                 coq-bignums │   55.90    55.30   1.08 │   153987715226    153372622287   0.40 │   206382243015    206405184548  -0.01 │  464348   464244   0.02 │   0    0    nan │
│                    coq-core │  107.15   105.91   1.17 │   296291937132    295799807704   0.17 │   458037038455    458119879183  -0.02 │  286044   286028   0.01 │   0    0    nan │
│              coq-coquelicot │   69.10    68.29   1.19 │   187771735232    186473222356   0.70 │   240000393046    239020192823   0.41 │  558356   557688   0.12 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
~~~

**Comment by SkySkimmer** [comment]
Seems to work ok

**Comment by herbelin** [comment]
@SkySkimmer: how did you arrive to this proposed change?

In any case, this looks like as much for @coq/typeclasses-maintainers than for @coq/stdlib-maintainers.

**Comment by SkySkimmer** [comment]
>@SkySkimmer: how did you arrive to this proposed change?

see posts in the issue

**Review by ppedrot — COMMENTED** [review]
Code change is fine, but can we get a changelog? This is the kind of change that can introduce weird behaviours and are hard to track.

**Comment by ppedrot** [comment]
I'll merge after the changelog is added, since there is no point in waiting more.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**
